### DISCUSSION TITLE
gcc@11 + CMake "cannot compile a simple test program"

### DISCUSSION THREAD

**Discussion by tcbrindle** [Homebrew/discussions#3755] [Linux] [answered]
Like many others, I was recently bitten by the [Github Actions changes](https://github.com/actions/runner-images/issues/6283), which broke Linux CI for my project.

Previously I installed GCC 11 via the straightforward `brew install gcc@11` in my CI script. When `brew` went missing from my $PATH, I used the Homebrew/actions/setup-homebrew@master action to re-add it. This seems to work okay, in that `brew install gcc@11` installs the compiler again, as requested.

Unfortunately I now have a another problem: when configuring CMake for my project with `CXX=g++-11` I get an error as follows:

```
-- Configuring incomplete, errors occurred!
  The C++ compiler
See also "/home/runner/work/libflux/build/CMakeFiles/CMakeOutput.log".

See also "/home/runner/work/libflux/build/CMakeFiles/CMakeError.log".

    "/home/linuxbrew/.linuxbrew/bin/g++-11"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/runner/work/libflux/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/make -f Makefile cmTC_c0a0f/fast && /usr/bin/make  -f CMakeFiles/cmTC_c0a0f.dir/build.make CMakeFiles/cmTC_c0a0f.dir/build
    make[1]: Entering directory '/home/runner/work/libflux/build/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_c0a0f.dir/testCXXCompiler.cxx.o
    /home/linuxbrew/.linuxbrew/bin/g++-11    -o CMakeFiles/cmTC_c0a0f.dir/testCXXCompiler.cxx.o -c /home/runner/work/libflux/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    Linking CXX executable cmTC_c0a0f
    /usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_c0a0f.dir/link.txt --verbose=1
    /home/linuxbrew/.linuxbrew/bin/g++-11 -rdynamic CMakeFiles/cmTC_c0a0f.dir/testCXXCompiler.cxx.o -o cmTC_c0a0f 
    /usr/bin/ld: cannot find -lm
```

I'm not really sure what's going on -- why can Homebrew-installed GCC now suddenly not find system libs? All was working fine a few days ago!

The GH Actions logs can be found [here](https://github.com/tcbrindle/libflux/actions/runs/3205204276/jobs/5237421385), and my (revised) CI script is [here](https://github.com/tcbrindle/libflux/blob/efe05aa29817cab49d20de26a492eb8bf69be5e1/.github/workflows/linux.yml). Please let me know if there are any more details which would be helpful :)

**Comment by carlocab** [comment]
Sorry, this was broken by recent changes that required making `glibc` keg-only. Homebrew/homebrew-core#111830 needs to be backported to `gcc@11`.

As a workaround in the meantime, please do
```
brew link --force glibc
```

CC @Homebrew/linux

**Reply by danielnachun** [reply]
I am going to backport the changes to all non-deprecated versions of GCC as they are trivial.

**Reply by danielnachun** [reply]
https://github.com/Homebrew/homebrew-core/pull/112555

**Comment by danielnachun** [comment, marked as answer]
https://github.com/Homebrew/homebrew-core/pull/112555 has been merged and should fix this.  Please try to run your workflow again and let us know if it works.

**Reply by tcbrindle** [reply]
Thank you!

I have just re-run my troublesome job and this time it gets to the build stage before erroring out with

```
[  2%] Building CXX object test/CMakeFiles/catch-main.dir/catch_main.cpp.o
as: unrecognized option '--gdwarf-5'
make[2]: *** [test/CMakeFiles/catch-main.dir/build.make:76: test/CMakeFiles/catch-main.dir/catch_main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:100: test/CMakeFiles/catch-main.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
Error: Process completed with exit code 2.
```

[Log](https://github.com/tcbrindle/libflux/actions/runs/3205204276/jobs/5241312165)

I'm not sure where the `--gdwarf-5` option is coming from, or why `as` doesn't like it.

**Reply by danielnachun** [reply]
You need to also have brewed `binutils` in your path, because the host `binutils` is too old.  It is keg-only so you'll need something like `PATH=$(brew --prefix)/opt/binutils/bin:$PATH` in your path.

**Reply by danielnachun** [reply]
Or alternatively, `brew link --force binutils`.

**Reply by tcbrindle** [reply]
Thank you, this fixed it 👍 

Are the binutils changes documented anywhere?
